### PR TITLE
Add support for test coverage test thresholds

### PR DESCRIPTION
**Pull request by redtangent**
## Description
Adds support for a test threshold for line and function coverage in the test script `basic-build-test.sh`, which will also cause the script to report an error if the threshold falls short, and to exit with an error code.

The idea is this can be used in the CI to indicate an error if the line or function coverage falls beneath a set threshold.

## Status
**READY**

## Requires Backporting
Yes  
Which branch?
`mbedtls-2.7` but not 2.1, as the script doesn't already exist there

## Migrations
NO

## Todos
- [ ] Tests
- [ ] Documentation
- [ ] Changelog updated
- [ ] Backported

### CODE REVIEW

File: tests/scripts/basic-build-test.sh [hunk 1]
```diff
@@ -42,6 +42,7 @@ fi
 : ${GNUTLS_SERV:="gnutls-serv"}
 : ${GNUTLS_LEGACY_CLI:="$GNUTLS_CLI"}
 : ${GNUTLS_LEGACY_SERV:="$GNUTLS_SERV"}
+: ${SEED:=1}
```
Comment (Patater) [inline tests/scripts/basic-build-test.sh]: What does the seed have to do with adding test coverage threshold testing? Seed related changes should go in a separate commit.

File: tests/scripts/basic-build-test.sh [hunk 2]
```diff
@@ -52,6 +53,51 @@ export GNUTLS_SERV="$GNUTLS_SERV"
...
+EXITCODE=0
+
+LINE_COV_THRESHOLD=0
+FN_COV_THRESHOLD=0
+
+print_usage() {
+    echo "Usage: $0 [options]"
+    printf "  -h|--help\tPrint this help.\n"
+    printf "     --seed\tInteger seed value to use for this test run (default 1)\n"
+    printf "     --line-cov-threshold\tLine coveraged test threshold (default 0)\n"
```
Comment (andresag01) [inline tests/scripts/basic-build-test.sh]: Minor: Shouldnt it be `coverage` instead of `coveraged`?

File: tests/scripts/basic-build-test.sh [hunk 3]
```diff
@@ -201,9 +246,18 @@ FUNCS_PERCENT=$((1000*$FUNCS_TESTED/$FUNCS_TOTAL))
...
 
 echo "Lines Tested       : $LINES_TESTED of $LINES_TOTAL $LINES_PERCENT%"
+if [ $(echo "$LINES_PERCENT < $LINE_COV_THRESHOLD" | bc -l) = "1" ]; then
+    echo "Test error! Line coverage was below test threshold $LINE_COV_THRESHOLD"
+    EXITCODE=1
+fi
+
 echo "Functions Tested   : $FUNCS_TESTED of $FUNCS_TOTAL $FUNCS_PERCENT%"
-echo
+if [ $(echo "$FUNCS_PERCENT < $FN_COV_THRESHOLD" | bc -l) = "1" ]; then
```
Comment (andresag01) [inline tests/scripts/basic-build-test.sh]: Minor: Is the `-l` option needed here for bc?

**Comment by andresag01** [comment]
@Patater: I think there is a CI that executes this test weekly and generate a report. I suppose this will affect that, but perhaps it would also be good to consider whether we want to run this a bit more often.

File: tests/scripts/basic-build-test.sh (hunk 1 above)
Comment (redtangent) [reply tests/scripts/basic-build-test.sh]: Fair point. I've split this into another commit, although doing this doesn't seem very common in the project.

File: tests/scripts/basic-build-test.sh (hunk 2 above)
Comment (redtangent) [reply tests/scripts/basic-build-test.sh]: Doh! Fixed.

File: tests/scripts/basic-build-test.sh (hunk 3 above)
Comment (andresag01) [reply tests/scripts/basic-build-test.sh]: Just to clarify: bc supports floats of arbitrary precision without the `-l` options, the problem is that the `scale` (the number of decimal digits after the point) is set to 0 by default. The effect of the `-l` option is to include the standard math libraries (which IMHO are not needed in this case) and to set the scale to 20 (according to POSIX at least). The script would work just as well if bc was also told to execute the statement `scale=20;` before the float arithmetic to set the precision to whatever is needed.

If the scale differences across different platforms is a concern, then I would recommend to manually set the precision manually (regardless of the `-l` option) so that we do not get slightly different results in different settings due to this problem...

Comment (redtangent) [reply tests/scripts/basic-build-test.sh]: I think we're looking at different documentation for `bc`, but now I look deeper, I can see this won't work for POSIX anyway, because apparently POSIX `bc` can't evaluate expressions outside of an `if`. I've removed the `-l` option, and given `basic-build-test.sh` only works on Linux today anyway, I guess this is good enough.

File: tests/scripts/basic-build-test.sh [hunk 4]
```diff
@@ -217,3 +271,5 @@ make clean
 if [ -f "$CONFIG_BAK" ]; then
     mv "$CONFIG_BAK" "$CONFIG_H"
 fi
+
+exit $EXITCODE
```
Comment (sbutcher-arm) [inline tests/scripts/basic-build-test.sh]: `$EXITCODE` should also indicate failure if any test fails - not just if the test coverage is below the set threshold.

Comment (redtangent) [reply tests/scripts/basic-build-test.sh]: Done!

File: tests/scripts/basic-build-test.sh [hunk 5]
```diff
@@ -185,6 +230,11 @@ echo "Total exec'd tests : $TOTAL_EXED"
 echo "Total avail tests  : $TOTAL_AVAIL"
 echo
 
+# Set exitcode if any tests failed
+if [ $(echo "$TOTAL_FAIL > 0" | bc) = "1" ]; then
```
Comment (andresag01) [inline tests/scripts/basic-build-test.sh]: I think that `$TOTAL_FAIL` should actually be an integer, so I suggest to use the regular arguments to the test command to avoid further use of `bc` as you said its problematic. For example:
```
if [ $TOTAL_FAIL -gt 0 ]; then
```

File: tests/scripts/basic-build-test.sh [hunk 6]
```diff
@@ -201,9 +251,18 @@ FUNCS_PERCENT=$((1000*$FUNCS_TESTED/$FUNCS_TOTAL))
 FUNCS_PERCENT="$(($FUNCS_PERCENT/10)).$(($FUNCS_PERCENT-($FUNCS_PERCENT/10)*10))"
 
 echo "Lines Tested       : $LINES_TESTED of $LINES_TOTAL $LINES_PERCENT%"
+if [ $(echo "$LINES_PERCENT < $LINE_COV_THRESHOLD" | bc) = "1" ]; then
```
Comment (andresag01) [inline tests/scripts/basic-build-test.sh]: So I checked the POSIX standard [here](http://pubs.opengroup.org/onlinepubs/000095399/utilities/bc.html) and for this to be portable the expression should be something like:
```sh
BC_CODE="scale = 20 ; if ($LINES_PERCENT < $LINE_COV_THRESHOLD) { 1 } ; if ($LINES_PERCENT >= $LINE_COV_THRESHOLD) { 0 } ;"
if [ $(echo "$BC_CODE" | bc) = "1" ]; then
```

**Comment by daverodgman** [comment]
This is manually checked regularly. It would be good to automatically check, but this PR is probably too old to merge & would be simpler to revisit from scratch.
